src/rmp/include/rmp/block_placement.h

```diff
@@ -0,0 +1,555 @@
+#ifndef BLOCK_PLACEMENT_H_
+#define BLOCK_PLACEMENT_H_
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <random>
+
+#include "rmp/shape_engine.h"
+
+namespace block_placement {
+    // definition of blocks:  include semi-soft blocks and soft blocks
+    class Block {
+        private:
+            float width_ = 0.0;
+            float height_ = 0.0;
+            float area_ = 0.0;
+            float x_ = 0.0;  // lx
+            float y_ = 0.0;  // ly
+            std::string name_;
+            bool is_soft_ = true;
+            int num_macro_ = 0;
+        
+            std::vector<std::pair<float, float> > aspect_ratio_;
+            std::vector<std::pair<float, float> > width_limit_; // This is in increasing order
+            std::vector<std::pair<float, float> > height_limit_; // This is in decreasing order
+
+            std::mt19937* generator_;
+            std::uniform_real_distribution<float>* distribution_;
+
+        public:
+            Block() {   };
```

Comment:
This will leave generator_ & distribution_ uninitialized, please set them to nullptr.